### DISCUSSION TITLE
Porting blink beyond Linux and BSD

### DISCUSSION THREAD

**Discussion by trungnt2910** [jart/blink#35] [Show and tell]
[This is **not** a feature request, just a discussion to see how far blink can go]

So I can see that `blink` has good support on Linux and BSD-based OSes (including Darwin/macOS).
But it can also be compiled and run on some other POSIX systems like the Emscripten WASM sandbox or POSIX.

Let's try it with [Haiku](http://haiku-os.org/), an ELF-based, (nearly) POSIX-compliant OS with a Sys-V ABI:

# Compiling
- Just follow the guides. It seems like the `Makefile` uses `cc` instead of `gcc` and Haiku doesn't make a `cc` symlink, so we have to `export CC=/boot/system/bin/gcc` before running `make`.
- There will be a few compile errors and missing symbols, but they're all similar to what one can meet when porting other programs to Haiku.

After everything's done, you will get:
![image](https://user-images.githubusercontent.com/57174311/213111972-c47677a0-77f8-4051-a0e0-70cf259853d2.png)

Quite fun, running `blink` - a _Linux_ emulator, on top of a _Haiku_ emulator, on top of _WSL1_, on **Windows**.

# Testing
After getting a basic binary done, the next goal is `make check` to complete.
### The first error I encountered:
![image](https://user-images.githubusercontent.com/57174311/213113264-ce4ef212-a656-4511-9e51-66dde1fa57b7.png)
(This time, I'm running `blink` on the real Haiku, not the emulator as my Haiku emulator is not mature enough to handle complex syscalls)
This can be easily fixed using a `UnXlat` for `ITIMER`.
### The next:
![image](https://user-images.githubusercontent.com/57174311/213113787-56b65987-faca-4612-a58e-91760507d31c.png)
I don't know about this yet, probably something different about the way Haiku handles signals, will investigate later.

**Update**: On the Hyclone emulator (Haiku emulator on Linux), I also managed to produce the same effect as on native Haiku. I also reproduced the same issue with `nanosleep_test`, which also seems to use Cosmopolitan platform detection.
![image](https://user-images.githubusercontent.com/57174311/213144468-94945385-1ebd-47d9-ba5e-f03631e100f3.png)

# Questions for this project
What's this project's view on community support for additional platforms?

While some projects welcome this, others don't allow upstreaming platform-specific patches as they introduce some more clutter in the codebase (it mostly doesn't affect the _compiled binary_ size though, as all changes are in the `Makefile`s or hidden under macros).

**Comment by trungnt2910** [comment]
It seems the second error is related to a function that _Cosmopolitan_ implements using platform detection. How is the platform detected by a `Cosmopolitan` binary in a `blink` sandbox? Is the recognized platform Linux (the one that `blink` emulates) or the real host platform itself?

**Reply by jart** [reply]
I just pushed my recent updates to Cosmopolitan. This change should illuminate how Cosmo is doing feature detection. https://github.com/jart/cosmopolitan/commit/006c44ff5d40ce279c8293a5a490617f3d5b65cd

**Comment by jart** [comment]
> What's this project's view on community support for additional platforms?
> 
> While some projects welcome this, others don't allow upstreaming platform-specific patches as they introduce some more clutter in the codebase (it mostly doesn't affect the compiled binary size though, as all changes are in the Makefiles or hidden under macros).

This codebase is designed to be cluttered. You're more than welcome to make contributions helping Blink support Haiku, since that means Cosmopolitan users will have access to another platform. Assuming modifying Blink to have better syscall emulation isn't possible, the hard thing you'll likely encounter is that currently only I can change the test binaries. So what I recommend doing is adding something like this to files such as `third_party/cosmo/cosmo.mk`:

```make
ifneq ($(HOST_OS), Haiku)
THIRD_PARTY_COSMO_TESTS += \
        o/$(MODE)/third_party/cosmo/2/setitimer_test.com.ok
endif
```

You'd have to refactor what's already in there into a variable though. Should be straightforward. After that I can possibly take a look and rebuild the binary.

Another option is to just copy cosmo's code for setitimer_test.c into the Blink codebase under `test/func/` and then delete the dependency on the prebuilt binary.

Don't worry too much about code size if it's something that leads to better syscall emulation for all platforms. For example, porting to Cygwin and FreeBSD has illuminated many cases where things like errno conditions really ought to be handled at the Blink layer, rather than punting it to the kernel. One example is checking access rights on file descriptors.

**Comment by trungnt2910** [comment]
Projects like Blink are also a good test case for other projects!

As my device does not run virtual machines quite well, my preferred workload is Windows -> WSL1 -> [Hyclone - Haiku Emulator on Linux](https://github.com/trungnt2910/hyclone) when porting apps to Haiku.

And Blink's `make check` has helped me discover dozens of Hyclone's bugs as well as a bug of WSL1's own!

**Reply by jart** [reply]
Awesome to hear that! Feedback like this is great incentive for me to write more system call tests.
